File: LeetCode/1277 Count Square Submatrices with All Ones.kt
class Solution {
    fun countSquares(matrix: Array<IntArray>): Int {
        val H = matrix.size
        val W = matrix.first().size
        val dp = ArrayList(List(H) { ArrayList(List(W) { 0 })})
        var ans = 0
        for (h in 0 until H) {
            for (w in 0 until W) {
                if (matrix[h][w] == 0) {
                    dp[h][w] = 0
                } else {
                    dp[h][w] = when {
                        h > 0 && w > 0 -> minOf(dp[h-1][w], dp[h-1][w-1], dp[h][w-1]) + 1
                        else -> 1
                    }
                    ans += dp[h][w]
                }
            }
        }
        return ans
    }
}

File: LeetCode/1277 Count Square Submatrices with All Ones.py
class Solution:
    def countSquares(self, matrix: List[List[int]]) -> int:
        H, W = len(matrix), len(matrix[0])
        S = [[0] * (W + 1) for _ in range(H + 1)]
        for h in range(H):
            for w in range(W):
                S[h + 1][w + 1] = S[h + 1][w] + S[h][w + 1] - S[h][w] + matrix[h][w]
        ans = 0
        for h in range(H):
            for w in range(W):
                if matrix[h][w] == 0:
                    continue
                n_h, n_w = h + 1, w + 1
                while n_h <= H and n_w <= W:
                    now_ones = S[n_h][n_w] - S[n_h][w] - S[h][n_w] + S[h][w]
                    if now_ones == (n_h - h) ** 2:
                        ans += 1
                    n_h += 1
                    n_w += 1
        return ans
